apps/web/api/config/axios.client.ts

```diff
@@ -0,0 +1,117 @@
+import "client-only";
+
+import { BACKEND_URL } from "@/lib/constants";
+import axios, { isAxiosError, type AxiosError } from "axios";
+import { jwtDecode } from "jwt-decode";
+import { AccessTokenCookie, RefreshTokenCookie } from "./cookies.client";
+import { setSession } from "./session.client";
+
+export const httpNoAuth = axios.create({
+	baseURL: BACKEND_URL,
+});
+
+export const http = axios.create({
+	baseURL: BACKEND_URL,
+});
+
+http.interceptors.request.use(async (config) => {
+	const accessToken = await getAccessToken();
+	if (accessToken) config.headers.set("Authorization", `Bearer ${accessToken}`);
+	return config;
+});
+
+// Add a response interceptor
+http.interceptors.response.use(
+	(response) => response,
+	async (error) => {
+		console.error("====", error?.response?.status, error?.response?.data);
+		const originalRequest = error.config;
+
+		// If the error status is 401 and there is no originalRequest._retry flag,
+		// it means the token has expired and we need to refresh it
+		if (isAccessExpired(error)) {
+			// `throw error` here will cause the "catch" block to run in following try-catch statement
+			if (originalRequest._retry) throw error;
+			console.info("refreshing the access token");
+			originalRequest._retry = true;
+
+			try {
+				const accessToken = await refreshAccessToken();
+				if (!accessToken) throw null;
+				originalRequest.headers.Authorization = `Bearer ${accessToken}`;
+				return axios(originalRequest);
+			} catch {
+				setSession(null);
+				window.location.reload();
+				return;
+			}
+		}
+
+		handleOtherFailures(error);
+	},
+);
+
+httpNoAuth.interceptors.response.use(
+	(response) => response,
+	(error) => {
+		console.error("====", error?.response?.status, error?.response?.data);
+		handleOtherFailures(error);
+	},
+);
+
+const refreshAccessToken = async () => {
+	const refreshToken = RefreshTokenCookie.get();
+	if (!refreshToken) return null;
+	try {
+		const accessToken = await getRefreshedToken({
+			refreshToken,
+		});
+		// TODO: test and check the SSR scenario, when a request is made in a RSC, cookies can't be updated
+		// Cookies can't be updated in the middleware as well but there should not be API calls in the middleware
+		// So the problem exists only in during SSR
+		setSession({
+			accessToken,
+			refreshToken,
+		});
+		return accessToken;
+	} catch {}
+	return null;
+};
+
+async function getRefreshedToken(data: { refreshToken: string }) {
+	const {
+		data: { access_token: accessToken },
+	} = await httpNoAuth.post<{ access_token: string }>("/tokens/refresh", data);
+	return accessToken as string;
+}
+
+function isAccessExpired(error: unknown) {
+	return (
+		isAxiosError(error) &&
+		typeof error.response?.data?.trace === "string" &&
+		error.response?.data?.trace.includes("JWT")
+	);
+}
```

Comment:
This is not how Django response to us. Check how @MohamedHamed12 implemented it.